pkg/controller/deployment/sync.go

```diff
@@ -510,15 +510,26 @@ func (dc *DeploymentController) updateDeploymentStatus(allRSs []*extensions.Repl
 
 // isScalingEvent checks whether the provided deployment has been updated with a scaling event
 // by looking at the desired-replicas annotation in the active replica sets of the deployment.
-func (dc *DeploymentController) isScalingEvent(d *extensions.Deployment) bool {
+func (dc *DeploymentController) isScalingEvent(d *extensions.Deployment) (bool, error) {
 	newRS, oldRSs, err := dc.getAllReplicaSetsAndSyncRevision(d, false)
 	if err != nil {
-		return false
+		return false, err
 	}
 	// If there is no new replica set matching this deployment and the deployment isn't paused
 	// then there is a new rollout that waits to happen
 	if newRS == nil && !d.Spec.Paused {
```

Comment:
A) and B) are different from C) because they involve two distinct updates
as opposed to a single update that happens in C) so it seems natural to let
scaling be handled as part of the rollout (you requested both at the same
time).

On Mon, Jul 18, 2016 at 6:54 PM, Phillip Wittrock <notifications@github.com>
wrote:

> In pkg/controller/deployment/sync.go
> <https://github.com/kubernetes/kubernetes/pull/28162#discussion_r71184893>
> :
>
> >  	newRS, oldRSs, err := dc.getAllReplicaSetsAndSyncRevision(d, false)
> >  	if err != nil {
> > -		return false
> > +		return false, err
> >  	}
> >  	// If there is no new replica set matching this deployment and the deployment isn't paused
> >  	// then there is a new rollout that waits to happen
> >  	if newRS == nil && !d.Spec.Paused {
>
> Thought about this some more over the weekend:
>
> New rs is created, it cannot scale up because the deployment has to scale
> down, old rs is scaled down, new rs is scaled up?
>
> If I am reading the code correctly, in this case the scaling down will
> happen in the rollingrollout could, not the scale code because a "scaling
> event" will never be seen as the labels have all been updated. The scaling
> will then happen as part of the rollout which behaves differently.
>
> Simple example: I have a router which uses host ports and the deployment's
> size equals the nodes in the cluster. I change my router not to use host
> ports anymore and I scale it up at the same time. If we try to scale before
> we rollout the new template then we simply won't be able to do so.
> Eventually I think we will rollout successfully but running scaling before
> rolling out is not optimal.
>
> This same problem exists for the proportional scaling as well right?
> Create a new RS, then immediately scale it. Do you agree that for the
> following 3 scenarios should behave similarly? Your current implementation
> makes C) behave differently than A) and B).
>
>    - A) Update image, then Scale
>       1. Create RS
>       2. Scale
>    - B) Scale, then update image
>       1. Scale
>       2. Create RS
>    - C) Scale and update image at the same time
>       1. Create RS + Scale
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/28162/files/b68dbfbffc725587ddfa28e82fa556f1b2deeaae#r71184893>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADuFf0VFGwu18X0OhxdEAm089AuAaTN1ks5qW6_LgaJpZM4JAOWg>
> .
>